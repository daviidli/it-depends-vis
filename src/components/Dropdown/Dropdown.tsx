import React, { useState, useRef, useEffect } from 'react';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import clsx from 'clsx';

import './Dropdown.scss';

export interface DropdownProps<T> {
	selectedValue: T,
	label: string,
	disabled: boolean,
	setSelectedValue: Function,
	values: Array<T>,
	className?: string
};

const Dropdown = (props: DropdownProps<string>) => {
	const { selectedValue, label, disabled, setSelectedValue, values, className } = props;

	const [labelWidth, setLabelWidth] = useState(0);
	const inputLabel = useRef<HTMLLabelElement>(null);

	useEffect(() => {
		setLabelWidth(inputLabel.current!.offsetWidth);
	}, []);

	return (
		<FormControl variant='outlined' className={clsx('dropdown', className)} disabled={disabled}>
			<InputLabel ref={inputLabel} id={label} className='label' margin='dense' color='secondary'>
				{ label }
			</InputLabel>
			<Select
				className='select'
				value={selectedValue}
				onChange={(e) => setSelectedValue(e.target.value)}
				margin='dense'
				labelId={label}
				labelWidth={labelWidth}
				color='secondary'
			>
				{ values.map((val, i) => (<MenuItem key={val + i} value={val}>{ val }</MenuItem>)) }
			</Select>
		</FormControl>
	);
}

export default Dropdown;
