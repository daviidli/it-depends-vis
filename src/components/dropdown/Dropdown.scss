@import '../../shared';

.dropdown {
	&.button {
		@include button-hover;
	}
}
