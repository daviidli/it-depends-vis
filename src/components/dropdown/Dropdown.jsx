import React from 'react';
import PropTypes from 'prop-types';
import {
	Popover, Button
} from 'antd';
import DropdownPanel from './DropdownPanel';
import './Dropdown.scss';

const Dropdown = ({
	icon, startCommit, endCommit, setStartCommit, setEndCommit, commits
}) => (
	<Popover
		className="dropdown"
		content={(
			<DropdownPanel {...{
				startCommit, endCommit, setStartCommit, setEndCommit, commits
			}}
			/>
		)}
		placement="bottomRight"
		trigger="click"
	>
		<div className="button"><Button type="default" shape="circle" size="large" icon={icon} /></div>
	</Popover>
);

Dropdown.propTypes = {
	icon: PropTypes.element.isRequired,
	startCommit: PropTypes.number.isRequired,
	endCommit: PropTypes.number.isRequired,
	setStartCommit: PropTypes.func.isRequired,
	setEndCommit: PropTypes.func.isRequired,
	commits: PropTypes.arrayOf(PropTypes.object).isRequired
};

export default Dropdown;
