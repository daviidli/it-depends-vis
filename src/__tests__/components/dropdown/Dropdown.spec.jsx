import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Dropdown from '../../../components/dropdown/Dropdown';

const defaultProps = {
	icon: <div />,
	startCommit: 0,
	endCommit: 6,
	setStartCommit: jest.fn(),
	setEndCommit: jest.fn(),
	commits: ['a', 'b', 'c', 'd', 'e', 'f']
};

const setup = (props = defaultProps) => {
	const wrapper = shallow(<Dropdown {...props} />);
	return { wrapper };
};

describe('Dropdown component', () => {
	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});
