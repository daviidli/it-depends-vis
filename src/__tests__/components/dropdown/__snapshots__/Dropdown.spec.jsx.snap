// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component should match snapshot 1`] = `
<Popover
  className="dropdown"
  content={
    <DropdownPanel
      commits={
        Array [
          "a",
          "b",
          "c",
          "d",
          "e",
          "f",
        ]
      }
      endCommit={6}
      setEndCommit={[MockFunction]}
      setStartCommit={[MockFunction]}
      startCommit={0}
    />
  }
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  overlayStyle={Object {}}
  placement="bottomRight"
  transitionName="zoom-big"
  trigger="click"
>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    icon={<div />}
    loading={false}
    shape="circle"
    size="large"
    type="default"
  />
</Popover>
`;
